File: spring-boot/src/main/java/springboot/security/CustomAuthenticationEntryPoint.java
package springboot.security;

import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import org.springframework.security.core.AuthenticationException;
import org.springframework.security.web.AuthenticationEntryPoint;
import springboot.common.enums.ErrorCode;
import springboot.domain.dto.AuthDto;

public class CustomAuthenticationEntryPoint implements AuthenticationEntryPoint {

    private final ObjectMapper objectMapper = new ObjectMapper();

    @Override
    public void commence(HttpServletRequest request, HttpServletResponse response,
        AuthenticationException authException) throws IOException, ServletException {
        String exception = (String) request.getAttribute("exception");
        ErrorCode unauthorized = ErrorCode.UNAUTHORIZED;
        if (exception.equals(unauthorized.getCode())) {
            this.setResponse(response, unauthorized);
        }
    }

    private void setResponse(HttpServletResponse response, ErrorCode errorCode) throws IOException {
        response.setContentType("application/json;charset=UTF-8");
        response.setStatus(errorCode.getHttpCode());
        AuthDto responseAuthorization = AuthDto.builder()
                                               .body(new ArrayList<>(List.of("Authentication failed.")))
                                               .build();
        String json = this.objectMapper.writeValueAsString(responseAuthorization);
        response.getWriter()
                .print(json);
    }
}

File: spring-boot-kt/src/main/kotlin/real/world/springbootkt/global/config/security/CustomAuthenticationEntryPoint.kt
package real.world.springbootkt.global.config.security

import com.fasterxml.jackson.databind.ObjectMapper
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.security.core.AuthenticationException
import org.springframework.security.web.AuthenticationEntryPoint
import real.world.springbootkt.global.common.BaseResponse
import real.world.springbootkt.global.common.ErrorCode

class CustomAuthenticationEntryPoint(private val objectMapper: ObjectMapper) : AuthenticationEntryPoint {
    override fun commence(
        request: HttpServletRequest, response: HttpServletResponse,
        authException: AuthenticationException?
    ) {
        val exception = request.getAttribute("exception") as? String
        val unauthorized: ErrorCode = ErrorCode.UNAUTHORIZED
        if (exception == unauthorized.code) {
            response.apply {
                this.contentType = "application/json;charset=UTF-8"
                this.status = unauthorized.httpCode
            }
            val json = objectMapper.writeValueAsString(BaseResponse(unauthorized, unauthorized.message))
            response.writer.print(json)
        }
    }
}